tests/scanner/scanners/retention_scanner_test.py

```diff
@@ -0,0 +1,396 @@
+# Copyright 2018 The Forseti Security Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tests for RetentionScanner."""
+
+import collections
+import json
+import unittest
+import mock
+from collections import namedtuple
+import tempfile
+
+from tests.scanner.test_data import fake_retention_scanner_data as frsd
+from tests.unittest_utils import ForsetiTestCase
+from tests.unittest_utils import get_datafile_path
+from google.cloud.forseti.common.gcp_type import organization
+from google.cloud.forseti.common.gcp_type import project
+from google.cloud.forseti.common.gcp_type import folder
+from google.cloud.forseti.scanner.audit import retention_rules_engine as rre
+from google.cloud.forseti.scanner.scanners import retention_scanner
+
+def get_expect_violation_item(res_map, bucket_id, rule_name, rule_index):
+    """Create violations for expected violation list"""
+    lifecycle_str = json.dumps(res_map.get(bucket_id).get_lifecycle_rule())
+
+    return rre.RuleViolation(
+        resource_id=bucket_id,
+        resource_name=u'buckets/'+bucket_id,
+        resource_type=res_map.get(bucket_id).type,
+        full_name=res_map.get(bucket_id).full_name,
+        rule_name=rule_name,
+        rule_index=rule_index,
+        violation_type=rre.VIOLATION_TYPE,
+        violation_data=lifecycle_str,
+        resource_data=res_map.get(bucket_id).data)
+
+yaml_str_multi_buckets_in_a_rule = """
+rules:
+  - name: multiple buckets in a single rule
+    applies_to:
+      - bucket
+    resource:
+      - type: bucket
+        resource_ids:
+          - fake-bucket-1
+          - fake-bucket-2
+          - fake-bucket-3
+    minimum_retention: 90
+
+"""
+
+
+def _mock_bucket_retention_multi_buckets_in_a_rule(_, resource_type):
```

Comment:
There is still A LOT of duplication between tests. Use closure to simplify:

Consider:

FakeBucketData = collections.namedtuple('FakeBucketData', ['bucket_id', 'project', 'age'])

```
def _get_mock_iter(fake_bucket_data):
    
    def _mock_iter(_, resource_type):
        if resource_type != 'bucket':
            raise ...
        resources = []
        for data in fake_bucket_data:
            data_creator = frsd.FakeBucketDataCreator(data.bucket_id, data.project)
          if data.age:
            data_creator.AddLifecycleDict(action='Delete', age=data.age)
          resources.append(data_creator.get_resource()) 
        return resources

    return _mock_iter

```